gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,219 @@
+      return new BatcherImpl<>(this);
+    }
+  }
+
+  public static <EntryT, EntryResultT, RequestT, ResponseT>
+      Builder<EntryT, EntryResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot add elements on a closed batcher.");
+
+    if (currentOpenBatch == null) {
+      currentOpenBatch = new Batch<>(batchingDescriptor.newRequestBuilder(prototype));
+    }
+
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+    currentOpenBatch.add(element, result);
+    return result;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() throws InterruptedException {
+    sendBatch();
+    awaitAllOutstandingBatches();
+  }
+
+  /** Sends accumulated elements asynchronously for batching. */
+  private void sendBatch() {
+    if (currentOpenBatch == null) {
+      return;
+    }
+    final Batch<ElementT, ElementResultT, RequestT> accumulatedBatch = currentOpenBatch;
+    currentOpenBatch = null;
+    numOfOutstandingBatches.incrementAndGet();
+
+    final ApiFuture<ResponseT> batchResponse =
+        callable.futureCall(accumulatedBatch.builder.build());
+
+    ApiFutures.addCallback(
+        batchResponse,
+        new ApiFutureCallback<ResponseT>() {
+          @Override
+          public void onSuccess(ResponseT response) {
+            try {
+              batchingDescriptor.splitResponse(response, accumulatedBatch.results);
+            } catch (Throwable ex) {
+              for (SettableApiFuture<ElementResultT> result : accumulatedBatch.results) {
+                result.setException(ex);
+              }
+            } finally {
+              onBatchCompletion();
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable throwable) {
+            try {
+              batchingDescriptor.splitException(throwable, accumulatedBatch.results);
+            } catch (Throwable ex) {
+              for (SettableApiFuture<ElementResultT> result : accumulatedBatch.results) {
+                result.setException(ex);
+              }
+            } finally {
+              onBatchCompletion();
+            }
+          }
+        },
+        directExecutor());
+  }
+
+  private void onBatchCompletion() {
+    if (numOfOutstandingBatches.decrementAndGet() == 0) {
+      synchronized (flushLock) {
+        flushLock.notifyAll();
+      }
+    }
+  }
+
+  private void awaitAllOutstandingBatches() throws InterruptedException {
+    while (numOfOutstandingBatches.get() > 0) {
+      synchronized (flushLock) {
+        flushLock.wait(DEFAULT_WAIT_TIME_MS);
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void close() throws InterruptedException {
+    isClosed = true;
+    flush();
+  }
+
+  /**
+   * This class represent one logical Batch. It accumulates all the elements and it's corresponding
```

Comment:
This class represents one logical batch. It accumulates all the elements and their corresponding future element results for one batch.